Add broadcasting support for `tf.where`

yongtang opened the pull request:
Adds where_v2 (which will be where in TF 2.0), which has numpy's broadcasting semantics.

This fix fixes #9284.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

drpngx commented:
@aselle WDYT?

ebrevdo commented on tensorflow/python/ops/array_ops.py:
```diff
@@ -2515,13 +2539,24 @@ def where(condition, x=None, y=None, name=None):
   has the same shape as `x` and `y`, then it chooses which element to copy from
   `x` and `y`.
 
+  If `broadcast` is True, then values of `x`, `y` and `condition` are
```
It sounds like the new behavior is backwards compatible?  Why hide it behind a flag if it doesn't break existing usage?

yongtang replied:
Thanks @ebrevdo. Initially I though the behaviors were different with respect to `broadcast`. Now think again it might be possible to not break the old behavior, while at the same time extend the `broadcast` 👍 . Let me take a look and update the PR.

yongtang commented:
@ebrevdo The PR has been updated with `broadcast` attribute removed. Please take a look.

tensorflowbutler commented:
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

ebrevdo commented on tensorflow/python/kernel_tests/where_op_test.py:
```diff
@@ -135,6 +135,34 @@ def testThreeArgument(self):
+    y = np.ones((7, 11))
+    np_val = np.where(f < 0, x, y)
+    with self.test_session(use_gpu=True):
+      tf_val = array_ops.where(constant_op.constant(f) < 0,
```
You should be able to use:
```
self.evaluate(f < 0, x, y)
```

same below.

ebrevdo commented on tensorflow/core/kernels/cwise_op_select.cc:
```diff
@@ -45,6 +45,58 @@ class SelectOp : public OpKernel {
+    BCast bcast(BCast::FromShape(cond->shape()),
+                BCast::FromShape(then->shape()));
+
+    if (bcast.IsValid()) {
```
add a comment about where this kicks in.  we currently have scalar broadcasting and primary-dimension vector broadcasting and i'd like to know if this will kick in in one of those existing cases -- because it may affect performance.

tensorflowbutler commented:
Nagging Assignee @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
Nagging Assignee @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

yongtang commented:
@ebrevdo The PR has been rebase to merge conflict.

However, after reviewing the code again, I realized that there is still one scenario where exiting `tf.where` is not compatible with `np.where` (below is the output before this PR):
```
$ python
Python 2.7.12 (default, Dec  4 2017, 14:50:18)
[GCC 5.4.0 20160609] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import tensorflow as tf
>>> import numpy as np
>>>
>>> x = np.arange(4)
>>> y = np.zeros((4, 4))
>>> z = np.ones((4, 4))
>>>
>>> np.where(x > 1, y, z)
array([[1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.]])
>>> v = tf.where(x > 1, y, z)
>>> tf.Session().run(v)
array([[1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.]])
>>>
```

As is shown from the above example, when x has the shape `(4)` and y/z has the shape `(4, 4)`, the broadcast orientation is different.

Because of that, the current PR will fail several test cases.

Unfortunately, I couldn't think of a way to make the proposed broadcasting changes in `tf.where` backward-compatible.

I am wondering maybe it would be better to name a new op (e.g., `tf.where_v2`) so that it is compatible with numpy and not breaking existing users?

carlthome commented:
Status?

velicue commented:
Any updates? Also as in the issue someone mentioned:

> There is also not full support for broadcasting in the other way, when condition is smaller than x or y; as indicated in the docs it only works if condition is a vector, but it could be extended to work with condition matching an arbitrary number of first dimensions of x and y (I'm not sure if this is considered also "broadcasting") and/or broadcasting singleton dimensions.

Would this also be implemented?

ebrevdo commented:
I think if you'd like the interface to exactly match np.where, it does make sense to create a new op + kernel, v2.  you would want to expose it in tf.contrib somewhere (not in core).

shoyer commented:
"If x and y are vectors of higher rank, then condition must be either a vector with size matching the first dimension of x, or must have the same shape as x."

Could we deprecate this behavior from `tf.where` (start issuing `FutureWarning`) and change it in some future API breaking release?

To ease the transition, we could safely add broadcasting support *when the total number of dimensions match* between all arguments, or add a function to contrib *temporarily* with the appropriate behavior.

Deviating from NumPy's broadcasting rules feels like a design mistake to me, and I suspect this will be a repeated source of confusion in the future.

tensorflowbutler commented:
Nagging Reviewer @aselle: It has been 36 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

yongtang commented:
Sorry for the delay. The PR has been updated. Now a new op `tf.contrib.framework.where` is exposed so that the broadcast rule follows numpy conventions. The original `tf.where` remains intact. Please take a look.

tensorflowbutler commented:
Nagging Reviewer @aselle: It has been 14 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

yongtang commented:
I rebased the PR to resolve the merge conflict though it looks like there are some build failures after that. Will take a look and update the PR shortly to fix the build.

charan223 commented:
Status guys?

yongtang commented:
The PR has been rebased with build error fixed. All test passed now. Sorry for the long wait.

yongtang commented:
@ebrevdo Please take a look and see if all issues have been addressed.

tensorflowbutler commented:
Nagging Assignee @ebrevdo: It has been 18 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
Nagging Assignee @ebrevdo: It has been 33 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
Nagging Assignee @ebrevdo: It has been 49 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

jhultman commented:
Would love to see this feature added

yongtang committed:
Add broadcasting support for `tf.where`
This fix tries to address the issue raised in 9284 where
there was no broadcasting support for `tf.where`. This
fix adds the support so that the behavior of `tf.where`
matches `np.where`.

This fix fixes 9284.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Update shape function for `tf.where` / `SelectOp`

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add template for BCastSelectFunctor

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add GPU support for where_v2

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add test case for broadcasting support of `where_v2`

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Define where_v2 in array_ops.py

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add broadcasting support for `tf.where`

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Fix `Experimental clang-format Check`

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Update api compatibility test

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Update api_def

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

ebrevdo commented:
+Martin for API review

josh11b commented:
Sorry for the delay!

For API review:

Changes to contrib do not require API review. That being said, we'd like to understand what is going on here. Are TF's broadcasting rules in general different than numpy or just for `where`? Did `where` have some custom broadcasting logic instead of using the same broadcasting library that other TF ops use? If so why is `where` special (e.g. is the difference that `where` takes three arguments and most of our ops that support broadcasting take two)?

Thanks!

aselle commented:
I think generally tf had very little broadcasting. As part of making things more numpy like, we changed select to be called where. Thus would reasonably expect broadcasting to happen. There are lots of corner cases of ops that don't support broadcasting in the same way. Another is tf.constant() which takes a shape but the tf behavior is when the shape doesn't match, it just pads with the last value where as in numpy constant will use broadcasting semantics to match the shape.

josh11b commented:
For API review:

Contrib is on the chopping block. If we want this behavior, we can't make it in v1.x because it is a backwards-incompatible change. Proposal: export this using: `@tf_export(v1="where_v2", v2="where")` and add an `@deprecation` to the current version of `where`.

ebrevdo commented:
I'd like to see how one would do batch major broadcasting with this version
of tf.where.  that's by far the most useful version and is the one we
currently have.

On Wed, Oct 3, 2018, 1:31 PM josh11b <notifications@github.com> wrote:

> For API review:
>
> Contrib is on the chopping block. If we want this behavior, we can't make
> it in v1.x because it is a backwards-incompatible change. Proposal: export
> this using: @tf_export(v1="where_v2", v2="where") and add an @deprecation
> to the current version of where.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/15982#issuecomment-426789882>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABtim2Lr2KqipOaBfD9UiDmNj2nQ7aZiks5uhR6FgaJpZM4RYgtJ>
> .
>

martinwicke commented:
I see... if we believe that the two broadcasting modes should be
accessible, we could alternative introduce an argument which switches the
behavior. In 1.x, the default must be the current behavior, in 2.0 we could
decide to switch the behavior, if we think consistency with numpy outweighs
the utility argument.

shoyer commented:
You could do "batch major broadcasting" the same way you do it for other
TensorFlow ops: you would explicitly expand dimensions first, e.g., with
tf.new_axis or tf.reshape.
On Thu, Oct 4, 2018 at 12:18 AM Martin Wicke <notifications@github.com>
wrote:

> I see... if we believe that the two broadcasting modes should be
> accessible, we could alternative introduce an argument which switches the
> behavior. In 1.x, the default must be the current behavior, in 2.0 we could
> decide to switch the behavior, if we think consistency with numpy outweighs
> the utility argument.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/15982#issuecomment-426822330>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABKS1nRkOn5--7LUNBXnir6wgveDDBRjks5uhTehgaJpZM4RYgtJ>
> .
>

ebrevdo commented:
This SGTM.  We'll have to update a number of internal users, especially
some RNN code, when moving to TF 2.0 if we make this the default.

On Wed, Oct 3, 2018 at 10:16 PM, Stephan Hoyer <notifications@github.com>
wrote:

> You could do "batch major broadcasting" the same way you do it for other
> TensorFlow ops: you would explicitly expand dimensions first, e.g., with
> tf.new_axis or tf.reshape.
> On Thu, Oct 4, 2018 at 12:18 AM Martin Wicke <notifications@github.com>
> wrote:
>
> > I see... if we believe that the two broadcasting modes should be
> > accessible, we could alternative introduce an argument which switches the
> > behavior. In 1.x, the default must be the current behavior, in 2.0 we
> could
> > decide to switch the behavior, if we think consistency with numpy
> outweighs
> > the utility argument.
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/tensorflow/tensorflow/pull/
> 15982#issuecomment-426822330>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/ABKS1nRkOn5--
> 7LUNBXnir6wgveDDBRjks5uhTehgaJpZM4RYgtJ>
>
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/15982#issuecomment-426888065>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABtimzeCkUUyVwmQ6CGi35ZmFjCEEJQQks5uhZm-gaJpZM4RYgtJ>
> .
>

ebrevdo commented:
can you add a test:

pred = np.array([1, 0, 0]).reshape((3, 1))
a = np.random.randn((3, 4))
b = np.random.randn((3, 4))

out = self.evaluate(pred, a, b)

assertAllClose(out, np.where(pred, a, b))

On Thu, Oct 4, 2018 at 10:17 AM, Eugene Brevdo <ebrevdo@google.com> wrote:

> This SGTM.  We'll have to update a number of internal users, especially
> some RNN code, when moving to TF 2.0 if we make this the default.
>
> On Wed, Oct 3, 2018 at 10:16 PM, Stephan Hoyer <notifications@github.com>
> wrote:
>
>> You could do "batch major broadcasting" the same way you do it for other
>> TensorFlow ops: you would explicitly expand dimensions first, e.g., with
>> tf.new_axis or tf.reshape.
>> On Thu, Oct 4, 2018 at 12:18 AM Martin Wicke <notifications@github.com>
>> wrote:
>>
>> > I see... if we believe that the two broadcasting modes should be
>> > accessible, we could alternative introduce an argument which switches
>> the
>> > behavior. In 1.x, the default must be the current behavior, in 2.0 we
>> could
>> > decide to switch the behavior, if we think consistency with numpy
>> outweighs
>> > the utility argument.
>> >
>> > —
>> > You are receiving this because you commented.
>> > Reply to this email directly, view it on GitHub
>> > <https://github.com/tensorflow/tensorflow/pull/15982#
>> issuecomment-426822330>,
>> > or mute the thread
>> > <https://github.com/notifications/unsubscribe-auth/
>> ABKS1nRkOn5--7LUNBXnir6wgveDDBRjks5uhTehgaJpZM4RYgtJ>
>>
>> > .
>> >
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/tensorflow/tensorflow/pull/15982#issuecomment-426888065>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ABtimzeCkUUyVwmQ6CGi35ZmFjCEEJQQks5uhZm-gaJpZM4RYgtJ>
>> .
>>
>
>

yongtang committed:
Add additional test case for tf.where_v2, based on review feedback

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@ebrevdo The PR has been updated with additional unit test. Please take a look.

tensorflowbutler commented:
Nagging Assignees @martinwicke, @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

yongtang committed:
Fix broken tests

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@ebrevdo @martinwicke Thanks for the help, the PR has been rebased to resolve the merge conflict. 

Please take a look and see if any additional change is needed.

tensorflowbutler commented:
Nagging Assignees @martinwicke, @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

martinwicke committed:
Merge branch 'master' into 9284-tf.where-broadcasting

martinwicke reviewed (commented):
this does not export where_v2, but it does export select_v2. Is this intended?

Can you make sure to export also to the v2 API (I think a golden update would do that), and make sure all functions that are supposed to be exported are? See @josh11b's comment: https://github.com/tensorflow/tensorflow/pull/15982#issuecomment-426789882.

yongtang committed:
Expose where_v2 as v1=["where_v2"], v2=["where"]

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add deprecation to tf.where

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Hide select_v2 API

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Update api goldens

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @martinwicke for the review. The PR has been updated with where_v2 exposed in @tf_export(v1=["where_v2"], v2=["where"]). The deprecation has also been added.

The issue with the `select_v2` was that I didn't add the `HIDDEN` in the api refs (the api systems have changed over time, so it is different from the time this PR is created). I have updated the api defs with `HIDDEN` added.

Please take a look and let me know if there are any further issues.

martinwicke commented on tensorflow/python/ops/array_ops.py:
```diff
@@ -3124,6 +3124,10 @@ def squeeze_v2(input, axis=None, name=None):
 
 
 @tf_export("where")
```
I'm surprised this works at all. This should be v1=["where"] (since there's another export for v2)

@annarev shouldn't this export conflict with the one below and throw an error in the api generation?

martinwicke commented:
Ah. v2= does nothing. @annarev, we should probably error out if we find a
named arg that's not useful.

Change the second export to tf_export("where", v1=["where_v2"]), then it
should work.

yongtang committed:
Update API to use @tf_export(v1=["where"]) for legacy `where`, and @tf_export("where", v1=["where_v2"]) for new `where`

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @martinwicke. The PR has been updated with `tf_export("where", v1=["where_v2"])` used for where_v2 and `tf_export(v1=["where"])` used for where (legacy) implementations. Please take a look.

yongtang commented:
@martinwicke @rthadur Wondering if there is any update on this PR? It would be good to see this PR in for v2, so that the broadcasting rule is consistent across different ops.

martinwicke commented:
@dksb Do you think it makes sense to scrub all PRs assigned to someone not on your team and assign one of your team to make sure they don't get lost like this one?

rthadur commented:
> @dksb Do you think it makes sense to scrub all PRs assigned to someone not on your team and assign one of your team to make sure they don't get lost like this one?

@yongtang @martinwicke @dksb  will take care of this PR. Feel free to assign me if you have such PR

martinwicke commented:
Oh, once I find the PR it's fine. It's more about the ones which I don't find. I usually find them only after the author pings me (or the PR), and that's usually after a long delay.

alextp commented:
Rasmus, can you review the kernel code? I'll review the API bits

alextp commented on tensorflow/python/ops/array_ops.py:
```diff
@@ -3177,6 +3181,46 @@ def where(condition, x=None, y=None, name=None):
 
 
+@tf_export("where", v1=["where_v2"])
+def where_v2(condition, x=None, y=None, name=None):
```
Is there ever a set of input shapes for which where and where_v2 both give answers but the answers are different?

I assume not, unless you're doing broadcast wrong.

In this case instead of a new API function you should go through forward compatibility, using the [compat.forward_compatible](https://github.com/tensorflow/tensorflow/blob/bd92ce08f8d5ae2dfa5cf47f39d86710215b7185/tensorflow/python/compat/compat.py#L34) function set to about a month in the future from now.

Search the tensorflow/python/ops directory for other uses of compat to see how to do this.

alextp replied:
Ah I see.

Then you also need to add a rule to [tf_upgrade_v2](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/compatibility/tf_upgrade_v2.py) to convert tf.where_v2 to tf.compat.v2.where as otherwise there can be a confusion when someone runs the upgrade script and tries to use their code in both v1 and v2.

alextp replied:
Since I wrote this message the tf v2 API has been frozen for the 1.14 release. This means we'll need to export this symbol as where_v2 both in tf1 and in tf2 :-/

yongtang commented:
Thanks @alextp for the review. There is a slight different in a corner case between v1 and v2. In v2, the broadcasting rule matches `np.where` and default broadcast rule in tensorflow. The v1 version of where is slightly different:
```python
import tensorflow as tf
import numpy as np
x = np.arange(4)
y = np.zeros((4, 4))
z = np.ones((4, 4))
# where_v2 matches np.where:
np.where(x > 1, y, z)

# Output:
array([[1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.]])

# where (v1):
v = tf.where(x > 1, y, z)
tf.Session().run(v)

# Output:
array([[1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.]])
```

ebrevdo commented:
Another way of looking at it is the old one broadcasts by batch and the new
one by depth (and more generally)

On Fri, Mar 29, 2019, 3:33 PM Yong Tang <notifications@github.com> wrote:

> Thanks @alextp <https://github.com/alextp> for the review. There is a
> slight different in a corner case between v1 and v2. In v2, the
> broadcasting rule matches np.where and default broadcast rule in
> tensorflow. The v1 version of where is slightly different:
>
> import tensorflow as tfimport numpy as np
> x = np.arange(4)
> y = np.zeros((4, 4))
> z = np.ones((4, 4))# where_v2 matches np.where:
> np.where(x > 1, y, z)
> # Output:
> array([[1., 1., 0., 0.],
>        [1., 1., 0., 0.],
>        [1., 1., 0., 0.],
>        [1., 1., 0., 0.]])
> # where (v1):
> v = tf.where(x > 1, y, z)
> tf.Session().run(v)
> # Output:
> array([[1., 1., 1., 1.],
>        [1., 1., 1., 1.],
>        [0., 0., 0., 0.],
>        [0., 0., 0., 0.]])
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/15982#issuecomment-478169199>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABtim7lthGHjADx_Jc2xysvFRmv05XG2ks5vbpTAgaJpZM4RYgtJ>
> .
>

yongtang committed:
Merge branch 'master' into 9284-tf.where-broadcasting

yongtang committed:
Update api compat and tf_upgrade_v2

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

alextp commented on tensorflow/tools/compatibility/renames_v2.py:
```diff
@@ -1515,6 +1515,10 @@
+    'tf.where':
+        'tf.compat.v1.where',
+    'tf.where_v2':
+        'tf.where',
```
It cannot be tf.where here because the output of the conversion script has to work in both v1 and v2. Instead this has to say tf.compat.v2.where

yongtang committed:
Rename tf.where_v2 to tf.compat.v2.where, as rename script has to

work in v1 and v2 so explicitly specifying tf.compat.v2.where
(from review comment).

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

alextp reviewed (commented):
Looks good to me. @rmlarsen , please approve if the kernel code looks good to you

rthadur commented:
> Looks good to me. @rmlarsen , please approve if the kernel code looks good to you

@rmlarsen gentle ping

ebrevdo commented:
@rmlarsen if we're gonna get this in we should do it soon.  TF-Agents relies on this broadcasting code and we're gonna have to make changes for tf2.0.

yongtang commented:
@martinwicke @rmlarsen @alextp Any chance to see this PR in 2.0? Really would like to see the consistency of broadcast rules across the board. (The 1.x tf.where broadcast rule is not compatible with the rest of the ops, and that could be some confusion for users).

rmlarsen commented:
Looking now.

rmlarsen commented on tensorflow/core/kernels/cwise_op_select.cc:
```diff
@@ -143,21 +143,138 @@ class SelectOp : public OpKernel {
+    // 2-ary broadcast.
+
+    // Combine `then` and `else`.
+    BCast elem_bcast(BCast::FromShape(then->shape()),
```
maybe call this then_else_bcast?

yongtang replied:
Thanks @rmlarsen, the name has been changed.

rmlarsen commented on tensorflow/core/kernels/cwise_op_select.cc:
```diff
@@ -324,10 +444,43 @@ struct BatchSelectFunctor<CPUDevice, T> {
 };
 
+template <typename T, int NDIMS>
+struct BCastSelectFunctor<CPUDevice, T, NDIMS> {
```
Why is this defined in multiple places? Can you just define this once in the header file and just template it on device type as well?

yongtang replied:
Thanks @rmlarsen, the PR has been updated with two definitions (one for CPU and one for SYCL) consolidated into one.

yongtang committed:
Rename elem_bcast to then_else_bcast, and remove duplicate template specification

based on review comment.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Fix GPU build failure

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@rmlarsen Thanks for the review. The PR has been updated. Please take a look and let me know if there are any issues.

martinwicke commented:
Since this has been going on for so long, I'd favor a cherry pick for this.

On Fri, May 3, 2019, 16:01 Alexandre Passos <notifications@github.com>
wrote:

> *@alextp* commented on this pull request.
> ------------------------------
>
> In tensorflow/python/ops/array_ops.py
> <https://github.com/tensorflow/tensorflow/pull/15982#discussion_r280951066>
> :
>
> > @@ -3177,6 +3181,46 @@ def where(condition, x=None, y=None, name=None):
>      raise ValueError("x and y must both be non-None or both be None.")
>
>
> +@tf_export("where", v1=["where_v2"])
> +def where_v2(condition, x=None, y=None, name=None):
>
> Since I wrote this message the tf v2 API has been frozen for the 1.14
> release. This means we'll need to export this symbol as where_v2 both in
> tf1 and in tf2 :-/
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/15982#discussion_r280951066>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEM57N3MKTMXHZBAVTHJQDPTS737ANCNFSM4ELCBNEQ>
> .
>

alextp commented:
We had to roll this back because the tests had no adequate coverage and missed things such as the broadcasting selectv2 op having no gradient defined for it.

yongtang commented:
@alextp sorry about that. I will take a look to add grad and resubmit the PR later.

brianwa84 commented:
I have a suggestion for the gradient, haven't tested it but maybe it gets you started.

```python

@ops.RegisterGradient("SelectV2")
def _SelectGrad(op, grad):
  c = op.inputs[0]
  x = op.inputs[1]
  y = op.inputs[2]
  zeros = array_ops.zeros([], dtype=grad.dtype.base_dtype)
  gx = array_ops.where_v2(c, grad, zeros)
  gx_shape = array_ops.shape(gx)
  x_shape = array_ops.shape(x)
  rankdiff_x = array_ops.rank(gx) - array_ops.rank(x)
  # Reduce away broadcasted leading dims.
  gx = math_ops.reduce_sum(gx, axis=math_ops.range(rankdiff_x))
  # Reduce but keep x's 1-valued dims which were broadcast.
  gx = math_ops.reduce_sum(
      gx, keepdims=1, axis=array_ops.where(grad_shape[rankdiff_x:] > x_shape))

  gy = array_ops.where_v2(c, zeros, grad)
  gy_shape = array_ops.shape(gy)
  y_shape = array_ops.shape(y)
  rankdiff_y = array_ops.rank(gy) - array_ops.rank(y)
  # Reduce away broadcasted leading dims.
  gy = math_ops.reduce_sum(gy, axis=math_ops.range(rankdiff_y))
  # Reduce but keep y's 1-valued dims which were broadcast.
  gy = math_ops.reduce_sum(
      gy, keepdims=1, axis=array_ops.where(grad_shape[rankdiff_y:] > y_shape))

  return (None, gx, gy)
```

tejaslodaya commented:
@yongtang any update on this? This was rolled back

brianwa84 commented:
It's supported in tf.compat.v2.where

On Wed, Dec 11, 2019, 7:43 AM Tejas Lodaya <notifications@github.com> wrote:

> @yongtang <https://github.com/yongtang> any update on this? This was
> rolled back
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/15982?email_source=notifications&email_token=AFJFSIY2TJ6NWWAGUSFZJDLQYDN5VA5CNFSM4ELCBNE2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEGS672A#issuecomment-564522984>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFJFSI3P5IYRBTGAJK6GOSLQYDN5VANCNFSM4ELCBNEQ>
> .
>

yongtang commented:
@tejaslodaya It was rolled back, then rolled forward (with the help from @brianwa84 for providing the gradient op 👍 ❤️ ). It is now available in 2.0.

tejaslodaya commented:
I tried and it works! For anyone coming to this PR, here's how you do it

Before (TF 1.x)-
```python
with tf.Session() as sess:  
    col = tf.convert_to_tensor([1,2,3,4,5,6,7,8,9,10,11,12])    
    print(tf.where(tf.math.greater(col, 10),
                  tf.zeros_like(col),
                  tf.ones_like(col)).eval())
```
> [1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 0. 0.]

After (TF 2.x)-
```python
import tensorflow as tf
col = [1,2,3,4,5,6,7,8,9,10,11,12]
print(tf.where(tf.math.greater(col, 10),
              tf.zeros([1]),
               tf.ones([1])))
```
> tf.Tensor([1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 0. 0.], shape=(12,), dtype=float32)


Notice, I had to do zeros_like and broadcast it to the shape of the column to make it work in 1.x

Thanks @yongtang , great work!
